desub-current/src/value/mod.rs

```diff
@@ -29,25 +29,142 @@ use serde::Deserialize;
 use std::convert::From;
 use std::fmt::Debug;
 
-/// [`Value`] holds a representation of some value that has been decoded.
+/// [`Value`] holds a representation of some value that has been decoded, as well as some arbitrary context.
 ///
 /// Not all SCALE encoded types have an similar-named value; for instance, the values corresponding to
 /// sequence, array and composite types can all be represented with [`Composite`]. Only enough information
 /// is preserved here to construct a valid value for any type that we know about, and be able to verify
 /// that a given value is compatible with some type (see the [`scale_info`] crate), if we have both.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Value<T> {
+	/// The shape and associated values for this Value
+	pub value: ValueDef<T>,
+	/// Some additional arbitrary context that can be associated with a value.
+	pub context: T,
+}
+
+impl Value<()> {
+	/// Create a new named composite value without additional context.
+	pub fn named_composite(values: Vec<(String, Value<()>)>) -> Value<()> {
+		Value { value: ValueDef::Composite(Composite::Named(values)), context: () }
+	}
+	/// Create a new unnamed composite value without additional context.
+	pub fn unnamed_composite(values: Vec<Value<()>>) -> Value<()> {
+		Value { value: ValueDef::Composite(Composite::Unnamed(values)), context: () }
+	}
+	/// Create a new variant value without additional context.
+	pub fn variant(name: String, values: Composite<()>) -> Value<()> {
+		Value { value: ValueDef::Variant(Variant { name, values }), context: () }
+	}
+	/// Create a new bit sequence value without additional context.
+	pub fn bit_sequence(bitseq: BitSequence) -> Value<()> {
+		Value { value: ValueDef::BitSequence(bitseq), context: () }
+	}
+	/// Create a new primitive value without additional context.
+	pub fn primitive(primitive: Primitive) -> Value<()> {
+		Value { value: ValueDef::Primitive(primitive), context: () }
+	}
+	/// Create a new `bool` value without additional context.
+	pub fn bool(val: bool) -> Value<()> {
+		Value { value: ValueDef::Primitive(Primitive::Bool(val)), context: () }
+	}
```

Comment:
I almost did make a macro for some of these! I think it'd end up looking like

`primitive!(char => Char)` 

(and maybe it would use `concat` to form a valid doc string attr to append on top

Reply:
Added some macros in https://github.com/paritytech/desub/pull/79/commits/b90fb7bcaedfce0597f51b8e6d98e537943a7107 :)